src/main/java/org/javarosa/xpath/analysis/XPathAnalyzer.java

```diff
@@ -0,0 +1,84 @@
+package org.javarosa.xpath.analysis;
+
+import org.javarosa.core.model.instance.TreeReference;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * An XPathAnalyzer is an object that can perform static analysis of any XPathAnalyzable
+ * (an XPathExpression or TreeReference) to ascertain specific semantic information about the
+ * raw content of the expression string itself
+ *
+ * @author Aliza Stone
+ */
+public abstract class XPathAnalyzer {
+
+    private TreeReference originalContextRef;
+    private TreeReference contextRef;
+    protected List<XPathAnalyzer> subAnalyzers;
+    protected boolean isSubAnalyzer;
+
+    public XPathAnalyzer() {
+        this.subAnalyzers = new ArrayList<>();
+    }
+
+    public XPathAnalyzer(TreeReference contextRef) {
+        this();
+        this.contextRef = contextRef;
+    }
+
+    public TreeReference getContextRef() {
+        return this.contextRef;
+    }
+
+    public TreeReference getOriginalContextRef() {
+        if (this.originalContextRef != null) {
+            return this.originalContextRef;
+        }
+        // Means that we only have 1 level of context
+        return this.contextRef;
+    }
+
+    public void doAnalysis(XPathAnalyzable analyzable) throws AnalysisInvalidException {
+        // So that the default behavior is to do nothing
+    }
+
+    // TODO: There should be special handling for references that contain "../" as well
+    public void doAnalysis(TreeReference ref) throws AnalysisInvalidException {
+        if (ref.getContextType() == TreeReference.CONTEXT_INHERITED) {
+            doAnalysisForRelativeTreeRef(ref);
+        } else if (ref.getContextType() == TreeReference.CONTEXT_ORIGINAL) {
+            doAnalysisForTreeRefWithCurrent(ref);
+        } else {
+            doNormalTreeRefAnalysis(ref);
+        }
+    }
+
+    public void doNormalTreeRefAnalysis(TreeReference treeReference)
+            throws AnalysisInvalidException {
+        // So that the default behavior is to do nothing
+    }
+
+    public void doAnalysisForTreeRefWithCurrent(TreeReference expressionWithContextTypeCurrent)
+            throws AnalysisInvalidException {
+        // So that the default behavior is to do nothing
+    }
+
+    public void doAnalysisForRelativeTreeRef(TreeReference expressionWithContextTypeRelative)
+            throws AnalysisInvalidException {
+        // So that the default behavior is to do nothing
+    }
+
+    public XPathAnalyzer spawnSubAnalyzer(TreeReference subContext) {
+        XPathAnalyzer subAnalyzer = initSameTypeAnalyzer();
+        subAnalyzer.isSubAnalyzer = true;
+        subAnalyzer.originalContextRef = this.getOriginalContextRef();
+        subAnalyzer.contextRef = subContext;
+        this.subAnalyzers.add(subAnalyzer);
```

Comment:
The method of keeping these in a recursive list seems a bit odd to me. Given that we have control over the internal lifecycle of the subanalyser, the pattern of

sub = SpawnSub
for (subcontexts())
  -- sub.analyse()
this.joinSub(sub)


feels like it would be a bit cleaner (we only maintain the subcontexts for their lifecycle) and not require as much context to be retained in the root object.

I think it's a pretty small change to shift to that, is there a specific advantage oyu were looking for from the current design?

Reply:
@ctsims So the problem I see with this suggestion is that the control flow really can't work the way that you're suggesting. Specifically, the `sub.analyze()` line can't be executed from within `XPathAnalyzer`, because this class has no knowledge of what the sub-analyzer is supposed to be analyzing or where; that information is known by `TreeReference.applyAndPropagateAnalyzer()` (which is where `spawnSubAnalyzer()` is called from). So in order to do what you're suggesting, `TreeReference.applyAndPropagateAnalyzer()` would have to have to make some call on the sub-analyzer _after_ the analysis was executed, telling it to propagate its results back up to its parent analyzer. That's definitely doable, but I don't think I see it as preferable, because then in the future, if other `XPathAnalyzable` classes have to generate sub-analyzers in the course of their analysis, they have to also remember to add this extra call. Whereas with the current setup, the aggregation happens automatically without any extra work.

Reply:
Yeah, I was assuming that the control flow logic for spawning and rejoining would need to be delegated back down to the same level as the current dispatch (IE: Inside of the xpath elements themselves currently). In the TreeReference implementation below the only difference would essentially be a line at the end of each iteration that says "originalAnalyzer.join(spawnedChildAnalyzer)" which knows how to put things back together.

I don't think it would change the future analyzer classes since all they'd be on the hook for would be implementing the "Join" method where they retrieve and process the results of their children. The valuable part to me of having that logic change is that it separates how we do the dispatch from what the analyzer is doing. 

Seems a bit cleaner to me, but also seems pretty easy to change in the future if there's a re-implementation that's needed, so no action needed if you don't think it's a valuable change at the moment.